File: core/src/main/java/com/zp4rker/iab/db/XYCoordPersister.java
package com.zp4rker.iab.db;

import com.j256.ormlite.field.FieldType;
import com.j256.ormlite.field.SqlType;
import com.j256.ormlite.field.types.StringType;
import com.zp4rker.iab.api.XYCoord;

import java.sql.SQLException;

public class XYCoordPersister extends StringType {
    public XYCoordPersister() {
        super(SqlType.STRING, new Class[]{XYCoord.class});
    }

    @Override
    public Object javaToSqlArg(FieldType fieldType, Object javaObject) throws SQLException {
        XYCoord coord = (XYCoord) javaObject;
        return coord.getX() + "," + coord.getY();
    }

    @Override
    public Object sqlArgToJava(FieldType fieldType, Object sqlArg, int columnPos) throws SQLException {
        String str = (String) sqlArg;
        if (!str.contains(",")) return null;

        String[] parts = str.split(",");
        int x = Integer.parseInt(parts[0]);
        int y = Integer.parseInt(parts[1]);
        return new XYCoord(x, y);
    }

    private static final XYCoordPersister singleton = new XYCoordPersister();

    public static XYCoordPersister getSingleton() {
        return singleton;
    }
}

File: src/main/kotlin/com/zp4rker/iab/db/XYCoordPersister.kt
package com.zp4rker.iab.db

import com.j256.ormlite.field.FieldType
import com.j256.ormlite.field.SqlType
import com.j256.ormlite.field.types.StringType
import com.zp4rker.iab.api.XYCoord

class XYCoordPersister : StringType(SqlType.STRING, arrayOf(XYCoord::class.java)) {
    override fun javaToSqlArg(fieldType: FieldType?, javaObject: Any?): Any {
        val xycoord = javaObject as XYCoord?
        return xycoord?.let { "${it.x},${it.y}" } ?: ""
    }

    override fun sqlArgToJava(fieldType: FieldType?, sqlArg: Any?, columnPos: Int): Any? {
        val str = sqlArg as String
        if (!str.contains(",")) return null

        val parts = str.split(",")
        val x = parts[0].toInt()
        val y = parts[1].toInt()
        return XYCoord(x, y)
    }

    companion object {
        private val singleton = XYCoordPersister()

        @JvmStatic
        fun getSingleton(): XYCoordPersister = singleton
    }
}